### PR TITLE
Add multi nic support

### PR DESCRIPTION
**Pull request by xulingqing**

### CODE REVIEW

**Comment by xulingqing** [comment]
@lambdai Could you take a look. 

ISTIO_EXCLUDE_INTERFACES  changes in https://github.com/istio/istio/pull/34832

File: scripts/vsphere/util.sh [hunk 1]
```diff
@@ -41,13 +41,16 @@ function fill_unspecified_context() {
   fi
 
   CLUSTER_NETWORK=$(kubectl -n istio-system get IstioOperator ${INSTALL_STATE} --output=json | jq -r '.spec.values.global.network')
+  if [[ -z ${VM_NETWORK} ]]; then
```
Comment (lambdai) [inline scripts/vsphere/util.sh:44]: ditto

File: scripts/vsphere/attach_vm [hunk 2]
```diff
@@ -160,12 +165,57 @@ write_config_context_vm() {
...
   rm ".attachConfig"
   cat ${OUTPUT_DIR}/vm-${VM_APP}/".context"
 }
 
+multinicsetup() {
+  echo "Starting Multi-nic setup"
+
+  IP_INFO = $(ssh -i "${KEY_FILE}" -o StrictHostKeyChecking=no root@${VM_IP} ip -j a )
+  IP_ADDR_VM = $(${IP_INFO} | jq -r '.[].addr_info[] | select(.family == "inet") | select(.scope == "global") |  .local' )
+  if [[ -z ${MANAGED_ADDR} ]]; then
```
Comment (lambdai) [inline scripts/vsphere/attach_vm:181]: -z "${MANAGED_ADDR}"

File: scripts/vsphere/attach_vm [hunk 3]
```diff
@@ -160,12 +165,57 @@ write_config_context_vm() {
...
+    # if the VM have more then 1 global IP, user need to specify the managed address
+    COUNT_IP_ADDR_VM = 0
+    for i in ${IP_ADDR_VM} ; do
+      COUNT_IP_ADDR_VM = $((COUNT_IP_ADDR_VM + 1))
+      if [[ ${#COUNT_IP_ADDR_VM} > 2 ]] ; then
+        # fatal "There is more than 1 global IP for the VM, user need to specify an address with --managed-address in attach_vm"
+        MANAGED_ADDR = $(${IP_INFO} | jq -r '.[] | select(.ifname=="${MANAGED_NIC}") | .addr_info[] | select(.family == "inet") | select(.scope == "global") | .local')
+      fi
+    done
+    MANAGED_ADDR = ${IP_ADDR_VM}
```
Comment (lambdai) [inline scripts/vsphere/attach_vm:191]: ditto

File: scripts/vsphere/attach_vm [hunk 4]
```diff
@@ -160,12 +165,57 @@ write_config_context_vm() {
...
+  else
+    # check if MANAGED_NIC and MANAGED_ADDR match for attachable VM
+    MANAGED_NIC_MAPPING_IP = $( ${IP_INFO} | jq -r '.[] | select(.ifname=="${MANAGED_NIC}") | .addr_info[] | select(.family == "inet") |  .local')
+    if [[ ${MANAGED_NIC_MAPPING_IP} != ${MANAGED_ADDR} ]] ; then
```
Comment (lambdai) [inline scripts/vsphere/attach_vm:195]: ditto

File: scripts/vsphere/attach_vm [hunk 5]
```diff
@@ -160,12 +165,57 @@ write_config_context_vm() {
...
+      fatal "MANAGED_NIC ${MANAGED_NIC} corresponding to IP ${MANAGED_NIC_MAPPING_IP} is different from user specified ${MANAGED_ADDR}."
+    fi
+  fi
+  echo "ISTIO_SVC_IP=${MANAGED_ADDR}" >> cluster.env
```
Comment (lambdai) [inline scripts/vsphere/attach_vm:199]: not sure

File: scripts/vsphere/attach_vm [hunk 6]
```diff
@@ -160,12 +165,57 @@ write_config_context_vm() {
...
+
+  # all NICs name exclude the $(managed-nic) name and "lo". Seperating with ","
+  IP_LINKS = $($(IP_INFO) | jq -r '.[].ifname')
+  EXCLUDE_NIC=""
+  for i in ${IP_LINKS} ; do
+    echo ${i}
+    if [[ ${i} != "lo" && ${i} != ${MANAGED_NIC} ]]; then
```
Comment (lambdai) [inline scripts/vsphere/attach_vm:206]: double quote

**Review by lambdai — COMMENTED** [review]
Please add double quote

File: scripts/vsphere/attach_vm (hunk 6 above)
Comment (xulingqing) [reply scripts/vsphere/attach_vm]: Thanks a lot!

File: scripts/vsphere/util.sh [hunk 7]
```diff
@@ -41,13 +41,16 @@ function fill_unspecified_context() {
   fi
 
   CLUSTER_NETWORK=$(kubectl -n istio-system get IstioOperator ${INSTALL_STATE} --output=json | jq -r '.spec.values.global.network')
+  if [[ -z "${VM_NETWORK}" ]]; then
+    VM_NETWORK="vmnet"
+    echo "VM Network      is not specified. Set VM Network as vmnet "
```
Comment (lambdai) [inline scripts/vsphere/util.sh:46]: Remove the extra spaces

File: scripts/vsphere/attach_vm [hunk 8]
```diff
@@ -160,12 +162,95 @@ write_config_context_vm() {
...
+# multi nic support
+multinicdetection() {
+  # cache all VM ip table info for later use
+  IP_INFO=$(ssh -i "${KEY_FILE}" -o StrictHostKeyChecking=no root@${VM_IP} ip -j a )
+
+  NICCOUNTSTRING=$(echo "${IP_INFO}" | jq -jr '.[].ifname| type')
+  NICSTRING=$(echo "${IP_INFO}" | jq -jr '.[]|" ",.ifname')
+
+  # Case 1: VM has only one nic. fill in MANAGED_ADDR and MANAGED_NIC if unspecified.
+  if [[ ${#NICCOUNTSTRING} == 12 ]]; then
```
Comment (lambdai) [inline scripts/vsphere/attach_vm:181]: can you explain what is the meaning of the magic number 12?

File: scripts/vsphere/attach_vm [hunk 9]
```diff
@@ -160,12 +162,95 @@ write_config_context_vm() {
...
+    INPUTNICIP=$(echo "${IP_INFO}"| jq -r '.[] | select(.ifname=="$(MANAGED_ADDR) | .addr_info[] | select(.family == "inet") | select(.scope == "global") | .local')
+
+    if [[ "${MANAGED_ADDR}" != "${VMNICIP}" || "${MANAGED_NIC}" != "${VMNICNAME}" ]]; then
+      # User might input a wrong addr or nic which doesn't match MANAGED_ADDR or MANAGED_NIC filled in above.
+      fatal "The VM has only one nic. nic name: ${VMNICNAME}, nic address: ${VMNICIP}. But user is specifying some other nic name or nic address."
+    elif [[ "${INPUTNICIP}" != "${VMNICIP}" ]]; then
+      # User might input a unmatched addr and nic
+      fatal "The VM has only one nic. nic name: ${VMNICNAME}, nic address: ${VMNICIP}. But user input for nic name and nic address is not matched."
+    fi
+  fi
```
Comment (lambdai) [inline scripts/vsphere/attach_vm:199]: Prefer `cannot find the nic ${MANAGED_NIC} or ${MANAGED_ADDR}`

File: scripts/vsphere/attach_vm [hunk 10]
```diff
@@ -160,12 +162,95 @@ write_config_context_vm() {
...
+    if [[ -z "${MANAGED_ADDR}" && -z "${MANAGED_NIC}" ]]; then
+      fatal "Detect undefined behavior. The VM is has more than one Nic. User need to specify managed nic name with one nic name out of ${NICSTRING}. Please specify with --managed-nic and --managed-address in the attaching process"
+    fi
+
+    if [[ -z "${MANAGED_ADDR}" ]]; then
+      # unspecified MANAGED_ADDR
+      MANAGED_ADDR=$(echo "${IP_INFO}"| jq -r ".[] | select(.ifname == \"${MANAGED_NIC}\") | .addr_info[] | select(.family == \"inet\" and .scope == \"global\") | .local")
+      # TODO: xulingqing(linggg) deal with case when a nic corresponding to multiple IPs
+      if [[ -z "${MANAGED_ADDR}" ]]; then
+        fatal "There is no nic with name ${MANAGED_NIC} on VM."
```
Comment (lambdai) [inline scripts/vsphere/attach_vm:213]: nit: Cannot find nic ${MANAGED_NIC}

**Review by lambdai — APPROVED** [review]
Generally LGTM modulo some nits.

File: scripts/vsphere/util.sh (hunk 7 above)
Comment (xulingqing) [reply scripts/vsphere/util.sh:46]: this is the alignment for the log.

File: scripts/vsphere/attach_vm (hunk 8 above)
Comment (xulingqing) [reply scripts/vsphere/attach_vm:181]: NICCOUNTSTRING is the count for how many nic on VM with concatenating "string" which is of length6. 
 if a VM has more than 2 nic like {lo ens192 ens224}, NICCOUNTSTRING="stringstringstring". User need to specify at least a --managed-nic or --managed-address.

File: scripts/vsphere/attach_vm (hunk 9 above)
Comment (xulingqing) [reply scripts/vsphere/attach_vm:199]: Is the point to not disclose that much info to user? Sure, changed.

**Pull request closed without merging**